src/main/java/com/flightstats/hub/app/GuiceToHK2Adapter.java

```diff
@@ -0,0 +1,91 @@
+package com.flightstats.hub.app;
+
+import com.google.inject.Injector;
+import com.google.inject.Key;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.glassfish.hk2.api.Factory;
+import org.glassfish.hk2.utilities.binding.AbstractBinder;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@Slf4j
+class GuiceToHK2Adapter extends AbstractBinder {
+
+    private final Injector injector;
+
+    GuiceToHK2Adapter(Injector injector) {
+        this.injector = injector;
+    }
+
+    @Override
+    protected void configure() {
+        injector.getBindings().forEach((key, value) -> {
```

Comment:
Will this handle singletons properly? I'm not at all familiar with HK2 and only know enough Guice for basic use cases. There's a way to check for singleton annotations in Guice, but I'm not sure how scoping or whatever works in HK2 binding/factories.

Reply:
Good question. I'll double check to be sure but I believe as long as the class is annotated with `javax.inject.Singleton` it should be ok. The problem is most of the code uses `com.google.inject.Singleton`. Because this will only effect new code it shouldn't be an issue as long as the dependencies of the new code are using the native injection stuff.

I'll take a look and see if we can check for the Google singleton annotation and bind it as such when setting up HK2. That would at least patch code that hasn't been migrated.

In practice I like to use the native injection stuff except where explicitly dealing with a library (e.g. a Guice module). Part of my larger DI branch involves migrating to the native injection tools.

Reply:
I believe guice'll find the native injection annotations, like you said. I think the bindings have some `isSingleton` method on them (or one of their properties) for guice. HK2 wasn't as clear in my brief doc reading. My concern was mostly legacy stuff where there's an annotation that's not being picked up here and we end up with tons of instances that hang around or were supposed to share state (ugh). Sounds like you've got it covered.

Reply:
...late on the follow-up so maybe not useful, but it looks like HK2 supports singleton scope with something like `bindFactory(...).to(..).in(Singleton.class)`

Reply:
Sweet! I was hoping I could just chain an `.singleton()` or something to that effect on the end. Thanks for looking that up.